src/terrain/terrain.js

```diff
@@ -938,40 +944,69 @@ export class Terrain extends Elevation {
         return this._style.order.some(isTransitioning);
     }
 
-    _clearRasterFadeFromRenderCache() {
-        let hasRasterSource = false;
-        for (const id in this._style._sourceCaches) {
-            if (this._style._sourceCaches[id]._source instanceof RasterTileSource) {
-                hasRasterSource = true;
+    _clearTilesFromRenderCache() {
+        let hasVectorOrRasterSource = false;
+        for (const source of this._style._getSources()) {
+            if (source instanceof VectorTileSource || source instanceof RasterTileSource) {
+                hasVectorOrRasterSource = true;
                 break;
             }
         }
-        if (!hasRasterSource) {
-            return;
-        }
 
-        // Check if any raster tile is in a fading state
+        if (!hasVectorOrRasterSource) return;
+
+        const clearSourceCaches = {};
         for (let i = 0; i < this._style.order.length; ++i) {
             const layer = this._style._layers[this._style.order[i]];
-            const isHidden = layer.isHidden(this.painter.transform.zoom);
             const sourceCache = this._style._getLayerSourceCache(layer);
-            if (layer.type !== 'raster' || isHidden || !sourceCache) { continue; }
+            if (!sourceCache || clearSourceCaches[sourceCache.id]) continue;
 
-            const rasterLayer = ((layer: any): RasterStyleLayer);
-            const fadeDuration = rasterLayer.paint.get('raster-fade-duration');
-            for (const proxy of this.proxyCoords) {
-                const proxiedCoords = this.proxyToSource[proxy.key][sourceCache.id];
-                const coords = ((proxiedCoords: any): Array<OverscaledTileID>);
-                if (!coords) { continue; }
-
-                for (const coord of coords) {
-                    const tile = sourceCache.getTile(coord);
-                    const parent = sourceCache.findLoadedParent(coord, 0);
-                    const fade = rasterFade(tile, parent, sourceCache, this.painter.transform, fadeDuration);
-                    const isFading = fade.opacity !== 1 || fade.mix !== 0;
-                    if (isFading) {
-                        this._clearRenderCacheForTile(sourceCache.id, coord);
-                    }
+            const isHidden = layer.isHidden(this.painter.transform.zoom);
+            if (isHidden) continue;
+
+            if (layer.type === 'line') {
+                clearSourceCaches[sourceCache.id] = true;
+                const lineLayer = ((layer: any): LineStyleLayer);
+                this._clearLineLayerFromRenderCache(lineLayer, sourceCache);
+            } else if (layer.type === 'raster') {
+                clearSourceCaches[sourceCache.id] = true;
+                const rasterLayer = ((layer: any): RasterStyleLayer);
+                this._clearRasterLayerFromRenderCache(rasterLayer, sourceCache);
```

Comment:
Should clearing tiles that are faded raster still be running on every frame? It makes sense to adopt a less frequent clear for line-widths as it's a small noticeable difference, but not clearing fading rasters might lead to flickering. 

Could you confirm with a satellite style with a long added [raster-fade-duration](https://docs.mapbox.com/mapbox-gl-js/style-spec/layers/#paint-raster-raster-fade-duration) to confirm that we're still ok on that?

Reply:
@karimnaaji I've separated the function that clears fading raster tiles, so it's now called on every frame as before.